src/Arcus.EventGrid/Contracts/EventGridEvent.cs

```diff
@@ -0,0 +1,109 @@
+            Guard.NotNullOrWhitespace(dataVersion, nameof(dataVersion));
+            Guard.NotNullOrWhitespace(eventType, nameof(eventType));
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EventGridEvent{TData}"/> class.
+        /// </summary>
+        /// <param name="id">The unique identifier of the event.</param>
+        /// <param name="data">The event data specific to the resource provider.</param>
+        /// <param name="dataVersion">The schema version of the data object. The publisher defines the schema version.</param>
+        /// <param name="eventType">The one of the registered event types for this event source.</param>
+        public EventGridEvent(string id, TData data, string dataVersion, string eventType)
+            : base(id, subject: null, data: data, eventType: eventType, eventTime: DateTime.UtcNow, dataVersion: dataVersion)
+        {
+            Guard.NotNullOrWhitespace(id, nameof(id));
+            Guard.NotNullOrWhitespace(dataVersion, nameof(dataVersion));
+            Guard.NotNullOrWhitespace(eventType, nameof(eventType));
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EventGridEvent{TData}"/> class.
+        /// </summary>
+        /// <param name="id">The unique identifier of the event.</param>
+        /// <param name="subject">The publisher-defined path to the event subject.</param>
+        /// <param name="data">The event data specific to the resource provider.</param>
+        /// <param name="dataVersion">The schema version of the data object. The publisher defines the schema version.</param>
+        /// <param name="eventType">The one of the registered event types for this event source.</param>
+        public EventGridEvent(string id, string subject, TData data, string dataVersion, string eventType)
+            : base(id, subject, data: data, eventType: eventType, eventTime: DateTime.UtcNow, dataVersion: dataVersion, metadataVersion: "1")
+        {
+            Guard.NotNullOrWhitespace(id, nameof(id));
+            Guard.NotNullOrWhitespace(subject, nameof(subject));
+            Guard.NotNullOrWhitespace(dataVersion, nameof(dataVersion));
+            Guard.NotNullOrWhitespace(eventType, nameof(eventType));
+        }
+
+        /// <summary>
+        ///     Event data specific to the resource provider.
+        /// </summary>
+        [JsonIgnore]
+        public new TData Data
+        {
+            get
+            {
+                if (base.Data is null)
+                {
+                    return default(TData);
+                }
+
+                if (base.Data is TData data)
+                {
+                    return data;
+                }
+
+                try
+                {
+                    return JObject.Parse(base.Data.ToString()).ToObject<TData>();
+                }
+                catch
+                {
+                    return default(TData);
```

Comment:
Yeah, or we could provide a method instead of a property to get the typed version of the data. It would though, have some minor changes in the clients who use them since our "old" `Event<>` class had also a prop named `Data`.

Reply:
Now that is an interesting point, I like the idea of exposing both types of data - Raw & typed.

`Data` would feel like the one where I get the typed version and would expect to get the raw one via `RawData`. However, given they already use `Data` we might want to move our stuff to `Payload` or something else. Any suggestions for names?

However, I do think that we should bubble the exception regardless of what went wrong, otherwise we are still covering up a potential real issue. Given we are going to add the raw field consumers can then use that in telemetry.

*PS: If we add the second fields some additional unit tests + integration tests would be good given we need to reassure that they are the same*

Reply:
* `Data`
* `GetData<T>` ?

Reply:
Given the object is typed we should not add a generic method imo and this feels more like a property than a method. How about `Data` & `Payload`?

> Payload: the actual information or message in transmitted data, as opposed to automatically generated metadata.

Reply:
Ok, I chagned it to `GetPayload()`

Reply:
I would bubble up the exception when deserializing the data fails, otherwise you're possibly hiding problems.

But, Isn't it a better idea to 'validate' the payload when setting this property ?  Or is that not possible ?

(But I feel like I'm commenting to outdated / unpushed code here ? )

Reply:
> Ok, I chagned it to `GetPayload()`

That is correct, but would prefer a property instead personally. Feels more natural to me and same approach as getting the raw information.

> I would bubble up the exception when deserializing the data fails, otherwise you're possibly hiding problems.
> 
> But, Isn't it a better idea to 'validate' the payload when setting this property ? Or is that not possible ?
> 
> (But I feel like I'm commenting to outdated / unpushed code here ? )

It's because the class has been changed since I've initially commented but your remarks are still applicable and share the same feeling.

In terms of validating I'm not really sure if that is possible, I think it's something we can explore but for now move to a dedicated issue maybe or use a follow-up PR so that we can merge this one?

Reply:
The properties are from  the parent type, so no; I don't think so. But feel free to experiment with the validation in a follow-up PR.